src/dpl/src/OptMirror.cpp

```diff
@@ -143,28 +159,177 @@ Opendp::findMirrorCandidates(NetBoxes &net_boxes,
+    vector<pair<dbNet*, Rect*>> updated_box;
+    int64_t delta = hpwlIncrement(inst, pt_x, true, net2box, updated_box);
+
+    if (delta <= 0) {
       mirror_count++;
+      inst->setLocationOrient(orient_my);
+      for (auto item : updated_box) {
+        net2box[item.first].reset(item.second->xMin(), item.second->yMin(),
+                                  item.second->xMax(), item.second->yMax());
+      }
     }
   }
   return mirror_count;
 }
 
+int64_t
+Opendp::hpwl(dbInst *inst)
+{
+  int64_t inst_hpwl = 0;
+  for (dbITerm *iterm : inst->getITerms()) {
+    dbNet *net = iterm->getNet();
+    if (net)
+      inst_hpwl += hpwl(net);
+  }
+  return inst_hpwl;
+}
+
+int64_t
+Opendp::hpwlIncrement(dbInst *inst, int pt_x, bool mirror,
+                      map<dbNet*, Rect> &net2box,
+                      vector<pair<dbNet*, Rect*>> &updated_box) {
+  int64_t delta = 0;
+
+  // Net to insterm mapping
+  map<dbNet*, vector<dbITerm *>> net2iterm;
+
+  for (dbITerm *iterm : inst->getITerms()) {
+    if (iterm->getNet() != NULL) {
+      dbNet *net = iterm->getNet();
+
+      if (net2iterm.find(net) == net2iterm.end()) {
+        net2iterm[net] = vector<dbITerm *>();
+        net2iterm[net].push_back(iterm);
+      } else {
+        net2iterm[net].push_back(iterm);
+      }
+    }
+  }
+
+  for ( auto &item : net2iterm ) {
+    dbNet *net = item.first;
+    if (!isSupply(net)) {
+      Rect *new_net_box = new Rect();
+      delta += hpwlIncrementHelper(inst, item.second, net, pt_x, mirror,
+                                  new_net_box, net2box[net]);
+      updated_box.push_back({net, new_net_box});
+    }
+  }
+  return delta;
+}
+
+int64_t
+Opendp::hpwlIncrementHelper(dbInst *inst, vector<dbITerm *> &iterms,
+                      dbNet *net, int pt_x, bool mirror, Rect *new_net_box,
+                      Rect netBox) {
+  int64_t net_hpwl = netBox.dx() + netBox.dy();
+  Rect iterm_box;
+  iterm_box.mergeInit();
+  int cellPtX, cellPtY;
+  inst->getLocation(cellPtX, cellPtY);
+  int cellWidth = inst->getMaster()->getWidth();
+
+  for (int i = 0; i < iterms.size(); i++) {
+    dbITerm *iterm = iterms[i];
+    int x, y;
+    if (iterm->getAvgXY(&x, &y)) {
+      Rect iterm_rect(x, y, x, y);
+      iterm_box.merge(iterm_rect);
+    } else {
+      // This clause is sort of worthless because getAvgXY prints
+      // a warning when it fails.
+      dbInst *inst = iterm->getInst();
+      odb::dbBox *inst_box = inst->getBBox();
+      int center_x = (inst_box->xMin() + inst_box->xMax()) / 2;
+      int center_y = (inst_box->yMin() + inst_box->yMax()) / 2;
+      Rect iterm_rect(center_x, center_y, center_x, center_y);
+      iterm_box.merge(iterm_rect);
+    }
+  }
+
+  // Based on current location check the pin location in the netbox
+  bool isInside = netBox.inside(iterm_box);
+
+  // Calculate the new pin delta displacement
+  int mirror_x = 0;
+  if (mirror) {
+    mirror_x = 2 * cellPtX + cellWidth - iterm_box.xMin() - iterm_box.xMax();
```

Comment:
Please explain the logic behind this calculation

Reply:
Iterm box llx location w.r.t cell's lower left point is  **iterm_box.xMin() - cellPtX**

Iterm box right most point distance from cell's right edge is **cellPtX + cellWidth - iterm_box.xMax()**

After flipping the pin displacement is **cellPtX + cellWidth - iterm_box.xMax() - (iterm_box.xMin() - cellPtX) = 2 * cellPtX + cellWidth - iterm_box.xMin() - iterm_box.xMax()**

Reply:
iterm_box.xMin() - iterm_box.xMax() is just iterm_box.dx()

Reply:
The expression is (- iterm_box.xMin() - iterm_box.xMax()) not (iterm_box.xMin() - iterm_box.xMax()) so it is not equal to iterm_box.dx()